crates/control/src/calibration_controller.rs

```diff
@@ -0,0 +1,285 @@
+use std::{time::Duration, vec};
+
+use color_eyre::Result;
+use serde::{Deserialize, Serialize};
+
+use calibration::{corrections::Corrections, measurement::Measurement, solve};
+use context_attribute::context;
+use coordinate_systems::Ground;
+use framework::{AdditionalOutput, MainOutput, PerceptionInput};
+use linear_algebra::{point, Point2};
+use types::{
+    calibration::{CalibrationCaptureResponse, CalibrationCommand},
+    camera_position::CameraPosition,
+    cycle_time::CycleTime,
+    field_dimensions::FieldDimensions,
+    primary_state::PrimaryState,
+};
+
+#[derive(Deserialize, Serialize)]
+pub struct CalibrationController {
+    pub current_look_at_index: usize,
+    pub current_calibration_state: InternalCalibrationState,
+    pub current_measurements: Vec<Measurement>,
+    pub last_calibration_corrections: Option<Corrections>,
+    pub look_at_list: Vec<(Point2<Ground>, CameraPosition)>,
+    pub last_capture_retries: u32,
+}
+
+#[context]
+pub struct CreationContext {}
+
+#[context]
+pub struct CycleContext {
+    primary_state: Input<PrimaryState, "primary_state">,
+    cycle_time: Input<CycleTime, "cycle_time">,
+    measurement_bottom: PerceptionInput<
+        Option<CalibrationCaptureResponse<Measurement>>,
+        "VisionBottom",
+        "calibration_measurement?",
+    >,
+    measurement_top: PerceptionInput<
+        Option<CalibrationCaptureResponse<Measurement>>,
+        "VisionTop",
+        "calibration_measurement?",
+    >,
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+    look_at_dispatch_wait_duration:
+        Parameter<Duration, "calibration_controller.look_at_dispatch_wait_duration">,
+    initial_to_calibration_stabilization_delay:
+        Parameter<Duration, "calibration_controller.initial_to_calibration_stabilization_delay">,
+    max_retries_per_capture: Parameter<u32, "calibration_controller.max_retries_per_capture">,
+
+    calibration_measurements: AdditionalOutput<Vec<Measurement>, "calibration_measurements">,
+    last_calibration_corrections:
+        AdditionalOutput<Option<Corrections>, "last_calibration_corrections">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub calibration_command: MainOutput<Option<CalibrationCommand>>,
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+pub enum InternalCalibrationState {
```

Comment:
@knoellle This is what I mentioned, is it okay to have it here or should I move it elsewhere in the file?

Reply:
Its fine here, but you can remove the `pub` and the "internal" prefix.

Reply:
Done, I realised I've unnecessarily made the fields public .